### ISSUE TITLE
Re-ordering topology file

### ISSUE THREAD

**Issue by Steboss** [ParmEd/ParmEd#469]
Hi!

I didn't know where exactly to post it, so I have tried here.

I have a problem with a topology, where the receptor comes first than the ligand. I need to swap the order, to ave first the ligand then the receptor, but I actually can't understand how to correctly handle some module of parmed

I was thinking to create a list for atoms properties and then use the command AmberParm.writeparm
Is the approach correct?
what could you suggest me?
thank you very much

Stefano

**Comment by hainm** [comment]
hi,

do you need to swap for simulation or just for visualization or other things? 

If just for other things (not care about the accuracy of parameters), you can just this

for example with `1D30` (DNA, then DAP ligand). If you want to move `DAP` to front:

```python
In [7]: p = pmd.download_PDB('1D30')

In [8]: [res.name for res in p.residues][:3]
Out[8]: ['DC', 'DG', 'DC']

In [9]: p  = p[':DAP'] + p['!:DAP']

In [10]: [res.name for res in p.residues][:3]
Out[10]: ['DAP', 'DC', 'DG']
```

In concise syntax

```python
In [11]: (p[':DAP'] + p['!:DAP']).save('new_pdb.pdb')
```

**Comment by hainm** [comment]
The main idea is to cut parm to smaller pieces and then reassemble them in new order (by using `+`)

**Comment by Steboss** [comment]
It's an excellent suggestion that could help 
I need for doing some simulations.
My starting point is a  prmtop and rst7 file with receptor - ligand order in.
I would like to have ligand-receptor  both in the topology and in the coordinate file

I was wondering if it's possible to do such a modification with parmed
Maybe create a class with all the parameters and coordinate for the receptor and one for the ligand (ions as well) and then male a  amberparm.writeparm with the order i want

**Comment by hainm** [comment]
@swails  might have different solution (like we can accurately swap amberparm by parmed).

But for your safety, you can use my method above to swap ligand and receptor, then save to pdb file. Then use `tleap` to rebuild prmtop and rst files.

```python
import parmed as pmd
# load parm7 and coords
p = pmd.load_file('1D30.parm7', '1D30.rst7')
# do swapping
p  = p[':DAP'] + p['!:DAP']
# save to pdb file for tleap
p.save('new_pdb.pdb')
```

then `tleap bla_bla`

**Comment by swails** [comment]
There's no need to do `tleap bla_bla`.  once you do `p = p[':DAP'] + p['!:DAP']`, you have a fully-functioning and parametrized object and can write a prmtop file directly.

**Comment by hainm** [comment]
ok. cool. 

@Steboss so the final code should be
```python
import parmed as pmd
# load parm7 and coords
p = pmd.load_file('1D30.parm7', '1D30.rst7')
# do swapping
p  = p[':DAP'] + p['!:DAP']
p.save('new.prmtop')
p.save('new.rst7')
```

**Comment by swails** [comment]
I have something way easier... hold on a sec.

**Comment by hainm** [comment]
@swails may be parmed website should have 1 page to have Q&A, or cookbook for small scripts so future users can look at it.

Few advantages

- clear demonstration about parmed's functionality

- ready to copy and paste

- show users about `parmed` idiom (like should using `load_file`, `save`, ... instead of calling different classes)

**Comment by Steboss** [comment]
@hainm @swails Thank you very much!
It's all clear
It would really be great if Parmed   could have a website for question and tutorial as well

Thank you very much,

Stefano

**Comment by swails** [comment]
OK, here's the recipe -- it took me a few seconds to try it out on one of my protein-ligand systems and make sure it worked and gave the same energies before and after:

> I was thinking to create a list for atoms properties and then use the command AmberParm.writeparm
> Is the approach correct?
> what could you suggest me?

Something even simpler:

```Python
import parmed as pmd

# Load in the topology file WITH COORDINATES
parm = pmd.load_file('your.prmtop', 'your.inpcrd')

# Swap the residues you want swapped. This code swaps residues 0 and 100
parm.residues[0], parm.residues[100] = parm.residues[100], parm.residues[0]
# You can do this more times if desired

# Here's the unfortunately tricky part.  You need to make sure that the atoms are re-sorted
# by residue
parm.atoms.sort(key=lambda x: x.residue.idx)

# And that's it.  Atoms resorted, everything is fine.  Now write out a prmtop and inpcrd
parm.save('reordered.prmtop')
parm.save('reordered.inpcrd')
```

The key aspects here:

1. You need to load coordinates -- the order of coordinates in the inpcrd must match the order of the atoms in the prmtop, so if you reorder the prmtop you need ParmEd to write a matching set of coordinates as well.

2. The residue and atom lists need to stay ordered in-sync, and Amber is often picky about atom ordering *within* a residue, so you can't change that.

The sort command will sort by residue ID, but the Python sort is stable so it is guaranteed to retain the same atom ordering *within* the residues.

**Comment by hainm** [comment]
Do you imply that `p[':DAP'] + p['!:DAP']` does not result correct energy?

**Comment by swails** [comment]
> Do you imply that p[':DAP'] + p['!:DAP'] does not result correct energy?

No, it should give the correct energy, but it's a lot slower and may not play quite as nicely with some weird potential things like 12-6-4 and NBFIXes.

**Comment by hainm** [comment]
> lot slower and may not play quite as nicely with some weird potential things like 12-6-4 and NBFIXes.

so should you add more tests for this? :P

**Comment by Steboss** [comment]
Does it change anything if the file are loaded with  amber.AmberParm rather than  load_file? and then re-write prmtop and rst7 with   writeParm and write_rst7 ?

**Comment by swails** [comment]
> Does it change anything if the file are loaded with amber.AmberParm rather than load_file? and then re-write prmtop and rst7 with writeParm and write_rst7 ?

No.  `load_file` has automatic file-type detection, so I usually use that to load any file ParmEd can read.  That way I never have to worry about the function call syntax or anything else with the input file I'm reading.

Likewise for `write_parm` (don't use `writeParm` -- it's gone in the current version) and `write_rst7` -- `.save()` is just more convenient for supported file types.  But if you're just as comfortable using the AmberParm-specific syntax (which is actually what `load_file` and `Structure.save` do under the hood), then by all means continue doing it :).

**Comment by Steboss** [comment]
Brilliant
It works perfectly :)

**Comment by Steboss** [comment]
test

**Comment by Steboss** [comment]
Hi
Here I am again 

I have update parmed and I got an error when I am using the code

```

import parmed as pmd

# Load in the topology file WITH COORDINATES
parm = pmd.load_file('your.prmtop', 'your.inpcrd')

# Swap the residues you want swapped. This code swaps residues 0 and 100
parm.residues[0], parm.residues[100] = parm.residues[100], parm.residues[0]
# You can do this more times if desired

# Here's the unfortunately tricky part.  You need to make sure that the atoms are re-sorted
# by residue
parm.atoms.sort(key=lambda x: x.residue.idx)

# And that's it.  Atoms resorted, everything is fine.  Now write out a prmtop and inpcrd
parm.save('reordered.prmtop')
parm.save('reordered.inpcrd')

```

Basically, when I am going to save the new topology I got this error:

```
/usr/local/lib/python2.7/dist-packages/parmed/amber/_amberparm.pyc in _set_owner(parm, owner_array, atm, mol_id)
   2255     """ Recursively sets ownership of given atom and all bonded partners """
   2256     parm.atoms[atm].marked = mol_id
-> 2257     for partner in parm.atoms[atm].bond_partners:
   2258         if not partner.marked:
   2259             owner_array.add(partner.idx)

/usr/local/lib/python2.7/dist-packages/parmed/topologyobjects.pyc in bond_partners(self)
    527             for c in p.children:
    528                 bp.add(c)
--> 529         return sorted(list(bp))
    530 
    531     @property

/usr/local/lib/python2.7/dist-packages/parmed/topologyobjects.pyc in __lt__(self, other)
    911 
    912     def __lt__(self, other):
--> 913         return self.idx < other.idx
    914 
    915     def __ge__(self, other):

RuntimeError: maximum recursion depth exceeded
```
I attached the input file I am using

Thank you 

[topology_coord.tar.gz](https://github.com/ParmEd/ParmEd/files/136635/topology_coord.tar.gz)

**Comment by swails** [comment]
I can reproduce the recursion depth exceeded error, but not in the same place yours is...  I'm completely at a loss here.  A bisect revealed that this was the last commit where your test worked: 92c253f029b029ca222305ded81bc73e8874e7cb

But I can't see anything in the next commit that would explain why this broke.  The only thing I can really suggest for now is to go back to that commit and use that version of ParmEd for your conversion.  I'll see if I can find what went wrong and why this infinite recursion is happening now, but I'm really stumped at the moment.

**Comment by Steboss** [comment]
Hi all
I  have a little update on this bug, which may focus on the right solution of the problem
I have a very small system (2 ions in a box) and I tried to swap them in order. 
All fine, this script works:
```
import parmed as pmd

# Load in the topology file WITH COORDINATES
parm = pmd.load_file('your.prmtop', 'your.inpcrd')

# Swap the residues you want swapped. This code swaps residues 0 and 100
parm.residues[0], parm.residues[100] = parm.residues[100], parm.residues[0]
# You can do this more times if desired

# Here's the unfortunately tricky part.  You need to make sure that the atoms are re-sorted
# by residue
parm.atoms.sort(key=lambda x: x.residue.idx)

# And that's it.  Atoms resorted, everything is fine.  Now write out a prmtop and inpcrd
parm.save('reordered.prmtop')
parm.save('reordered.inpcrd')
```
giving me no error

The error comes after, when I try to re-load the new topology in parmed and I have this error:
```
/usr/local/lib/python2.7/dist-packages/ParmEd-2.7.2_47.gb3336f3-py2.7-linux-x86_64.egg/parmed/amber/_amberparm.pyc in __init__(self, prm_name, xyz, box, rst7_name)
    180             self.crdname = xyz
    181         if prm_name is not None:
--> 182             self.initialize_topology(xyz, box)
    183 
    184     #===================================================

/usr/local/lib/python2.7/dist-packages/ParmEd-2.7.2_47.gb3336f3-py2.7-linux-x86_64.egg/parmed/amber/_amberparm.pyc in initialize_topology(self, xyz, box)
    211                 self.combining_rule = 'lorentz'
    212         # Instantiate the Structure data structures
--> 213         self.load_structure()
    214 
    215         if isinstance(xyz, string_types):

/usr/local/lib/python2.7/dist-packages/ParmEd-2.7.2_47.gb3336f3-py2.7-linux-x86_64.egg/parmed/amber/_amberparm.pyc in load_structure(self)
    503         self._check_section_lengths()
    504         self._load_atoms_and_residues()
--> 505         self.load_atom_info()
    506         self._load_bond_info()
    507         self._load_angle_info()

/usr/local/lib/python2.7/dist-packages/ParmEd-2.7.2_47.gb3336f3-py2.7-linux-x86_64.egg/parmed/amber/_amberparm.pyc in load_atom_info(self)
    554             anum = [-1 for atom in self.atoms]
    555         for i, atom in enumerate(self.atoms):
--> 556             atom.name = anam[i]
    557             atom.charge = chg[i]
    558             atom.mass = mass[i]

IndexError: list index out of range
```
Here I attached the input files I have used and my script to reproduce the error

Thanks very much

Stefano

[reproduce_name_error.zip](https://github.com/ParmEd/ParmEd/files/878808/reproduce_name_error.zip)

**Comment by craldaz** [comment]
Hi this is a very old thread, but I tried to reorder my own topology using the first method above,

**Comment by craldaz** [comment]
Hi this is an older thread, bu  when i follow the above procedure (e.g. p  = p[':DAP'] + p['!:DAP'])
All the bonds and other parameters between the residues disappears. 

How can I reorder a topology without deleting parameters between residues?

**Comment by craldaz** [comment]
The bonds and parameters between residues are not removed when performing the second procedure. However, I can now confirm that there is still an error with reloading the topology file with parmed 3.2.0 

 ```
 line 554, in load_atom_info
    atom.name = anam[i]
IndexError: list index out of range
```
